examples/computer-vision-server-side-streaming/readme.md

```diff
@@ -0,0 +1,89 @@
+
+Anyway, I finally managed to sort out the approach here where the server initiates the connection - and it seems to work. (Well, mostly.) This is actually kinda cool for other purposes too. But please note that I'm not an expert, and I only figured it out by brute force - so it might not be the best approach.
+
+## Parallelism and such
+
+So, I spent a lot of time on this and unless you can see where I'm obviously wrong (which is highly like as I'm not an expert in this), I'd suggest sticking with how the code is set up now. If you want to delve in, things to consider:
+
+- You don't want to block the event loop, so you want your compute intensive stuff (video decoding and the model running) in different processes, as they are here.
+- Since they're in different processes (not threads) you need to share the frames etc. between them. Shared memory as it's done here seems to be the easiest (remembering the numpy arrays are big and you don't want to throw them round lightly at 25FPS), though there are other approaches that'll work too, including with zero-copy e.g. ZeroMQ.
+- It'd be really nice to avoid asyncio, as it makes things a lot more complicated, e.g. if you want to use it in non-async code like a Flask app. I'm not enough of an expert, but I suspect wrapping all the few async aiortc calls in `loop.run_until_complete` might be the easiest way to integrate things.
+
+# A challenge ...
+
+So, I want to be able to have a low-bandwidth solution *and* guaranteed synchronicity between my frames and the inference results. (I know, it won't be low latency as it'll be delayed by at least the model runtime, but hey.) The first constraint means video ... and when you display video the modern way, it's very hard to know the actual frame that's currently being displayed - i.e. there's no frame number attribute, which would otherwise make this easy, as then you could say "OK, frame 123 arrived so let's draw the model results on frame 123". I see two potential approaches, and I'd love if someone implemented one (or another you know of) ...
+
+1. Use the WebRTC stats which, as you've seen, has the number of frames received. This is effectively the frame number, right? My main concern here is getting it in sync (i.e. knowing exactly which frame is the first that gets displayed and increments the count to one), and how poor connection affects things (e.g. if some frames are dropped, and the count not incremented, then everything will be out). And, I guess, how it might vary from browser-to-browser, etc.
+2. Do some steganography! Basically, encode the frame number in the pixels of the image, and then decode it on the frontend. The challenges on this front:
```

Comment:
Is it possible to interleave metadata or other non-image "chunk" data in the encoded video stream? (I'm rusty on this, but IIRC it should be possible.)

Reply:
Great question! In my experience, I haven't seen anything that works easily. The trick is, like you say, having access to it when you write *and* being able to get it out when you render it in the browser. I think I've seen `side_data` before but never looked in detail, so it definitely could work. (Likewise I've seen some things done with the subtitle.) Anyway - I would be *very* happy if you could figure it out - that'd be awesome!

Reply:
Oooh, a colleague just found this https://github.com/w3c/webrtc-insertable-streams/blob/master/explainer.md Which might be the trick for this last bit. I'll update now.